### PR TITLE
build: cache generated files across builds

### PR DESCRIPTION
**Pull request by zzmp**
## Description

Splits the `yarn prepare` step so that its components can utilize caching in CI builds.
Updates the setup action workflow to cache compiled contracts, generated graphql schema, extracted translations, and compiled localizations across builds if they have not changed.

`yarn prepare` will continue to work locally, but it's intention is now to be used to setup local development. CI, which was partially bottle-necked by these cacheable steps, will now call directly into `yarn prepare`s component steps.

### QA (ie manual testing)
- [ ] Integration/E2E test

### CODE REVIEW

File: lingui.config.ts [hunk 1]
```diff
@@ -1,8 +1,60 @@
+import { default as babelExtractor } from '@lingui/cli/api/extractors/babel'
+import { createHash } from 'crypto'
+import { mkdirSync, readFileSync, writeFileSync } from 'fs'
+import * as path from 'path'
+import * as pkgUp from 'pkg-up'
```
Comment (JFrankfurt) [inline lingui.config.ts:5]: you might consider find-up instead, as it's the same functionality with more usage https://www.npmjs.com/package/find-up

File: lingui.config.ts [hunk 2]
```diff
@@ -1,8 +1,60 @@
...
+    const root = path.dirname(pkg)
+
+    const filePath = path.join(root, filename)
+    const file = readFileSync(filePath)
+    const hash = createHash('sha256').update(file).digest('hex')
+
+    const cacheRoot = path.join(root, 'node_modules/.cache/lingui')
+    mkdirSync(cacheRoot, { recursive: true })
+    const cachePath = path.join(cacheRoot, filename.replace(/\//g, '-'))
+    if (!process.argv.includes('--clean')) {
```
Comment (JFrankfurt) [inline lingui.config.ts:32]: is clean a new arg?

Comment (JFrankfurt) [reply lingui.config.ts:32]: ah, I see it used in the script 👍

File: lingui.config.ts [hunk 3]
```diff
@@ -1,8 +1,60 @@
...
+      try {
+        const cache = readFileSync(cachePath, 'utf8')
+        if (cache === hash) return
+      } catch (e) {
+        // It should not be considered an error if there is no cache file.
+      }
+    }
```
Comment (JFrankfurt) [inline lingui.config.ts:39]: async readFile might read better here

**Review by JFrankfurt — APPROVED** [review]

File: lingui.config.ts (hunk 2 above)
Comment (zzmp) [reply lingui.config.ts:32]: deserves a comment, though

File: lingui.config.ts (hunk 3 above)
Comment (zzmp) [reply lingui.config.ts:39]: I think this function still needs to return synchronously, though

File: lingui.config.ts (hunk 1 above)
Comment (zzmp) [reply lingui.config.ts:5]: `pkgUp` is used internally by `lingui`, so I used it here so I could match the code from lingui's own extractors

Comment (cmcewen) [reply lingui.config.ts:5]: could you add this as a comment? useful info

File: lingui.config.ts [hunk 4]
```diff
@@ -1,8 +1,63 @@
+import { default as babelExtractor } from '@lingui/cli/api/extractors/babel'
```
Comment (cmcewen) [inline lingui.config.ts:1]: is this something you could upstream to this package?

**Review by cmcewen — APPROVED** [review]

File: lingui.config.ts (hunk 4 above)
Comment (zzmp) [reply lingui.config.ts:1]: yes, but it'd need to be a little more fully-featured.
but if we made time for it I'd love to do it.

**Review by JFrankfurt — APPROVED** [review]

**Pull request merged**